core/headerchain.go

```diff
@@ -192,6 +199,323 @@ func NewHeaderChain(db ethdb.Database, engine consensus.Engine, pEtxsRollupFetch
 	return hc, nil
 }
 
+// GetEngineForPowID returns the consensus engine for the given PowID
+func (hc *HeaderChain) GetEngineForPowID(powID types.PowID) consensus.Engine {
+	// Engine mapping:
+	// engine[0] = Progpow
+	// engine[1] = Kawpow
+	// engine[2] = SHA256d
+	switch powID {
+	case types.Progpow:
+		if len(hc.engine) > 0 {
+			return hc.engine[0]
+		}
+	case types.Kawpow:
+		if len(hc.engine) > 1 {
+			return hc.engine[1]
+		}
+	case types.SHA_BTC, types.SHA_BCH:
+		if len(hc.engine) > 2 {
+			return hc.engine[2]
+		}
+	}
+	// Default to first engine if not found
+	if len(hc.engine) > 0 {
+		return hc.engine[0]
+	}
+	return nil
+}
+
+// GetEngineForHeader returns the consensus engine for the given header
+func (hc *HeaderChain) GetEngineForHeader(header *types.WorkObjectHeader) consensus.Engine {
```

Comment:
Should we be getting the engine based on the powID instead of if auxpow is nil?

Reply:
pow id is not defined for progpow, so that wouldnt work

Reply:
but do we need GetEngineForHeader for SHA/Scrypt?